### DISCUSSION TITLE
How to make a camera in PIXI? How do you make side-scroller platformer games? PIXI-viewport - is a waste of performance in this case

### DISCUSSION THREAD

**Discussion by ProgrammingLife** [pixijs/pixijs#9490] [Q&A]
Let's say we have some basic side-scroller platform game or any other kind of side-scroller game with a lot of objects those are not in the viewport at this time. There are might be: tiles, enemies, objects, any other stuff.

The first approach is to add all those objects into some rootContainer. It's maybe `pixi-viewport `container, why not.
And we get a huge loss of performance since you should move all those objects using CPU every frame. Sounds like not the best solution for a side-scroller game.

So what to use else? Is there some cameras in PIXI those can clip off-screen objects?
Maybe `pixi-projection` plugin is coming into place here? Or maybe even `ParticleContainer`? But it's also a waste of CPU performance when moving a lot of off-screen objects every tick.

So, what are your thoughts, guys?

**Comment by bigtimebuddy** [comment]
I cannot speak to pixi-viewport because it's a 3rd party plugin. You should consider a discussion on that repo to see if there's something specific to the plugin creating issues.

Generally: it's very important that you cull your offscreen objects (set `object.visible = false`) regardless of how you're rendering your scene. Pixi has a property called `cullable` that can be used to automatically cull objects, though there can be more performant ways to do this. Doing this simple step will help. Reducing textures to not break batching, pooling offscreen objects to reduce your memory, etc can all help too.

Also, please checkout these other [performance tips](https://pixijs.io/guides/production/performance-tips.html).

**Reply by ProgrammingLife** [reply]
Thanks for the reply. But I'm not talking about culling, I'm talking about moving a lot of objects every frame instead of when you have a camera you don't need to reposition a lot of objects. Let's say you have a root container that contains 5000 objects: tiles, static objects those don't need to move. You move the character, so you need to move the container underneath, so you change x, y for all the 5000 objects in the container every tick even if they are invisible and offscreen at this time. You can't avoid it because it's a side-scroller game. 

Other option is to use camera projection. In this case, you don't need to move all the objects. Instead of it you can clip all invisible objects and render visible only without moving the root container or objects inside it.

**Reply by ProgrammingLife** [reply]
Maybe in PIXI v8 we will see an ability to use a true projection camera? 😀
Like it works in Phaser, Unity, etc...
It's really important thing. Specially when you make things like side-scroller platformer otherwise PIXI is not the good choice for such games with moving camera. We need a camera support! 😀

**Reply by fusionlove** [reply]
zoom and translation are tractable transforms to apply to your objects yourself
